### PR TITLE
Release v0.3.0

### PR DESCRIPTION
**Pull request by ipochi**
Signed-off-by: Imran Pochi <imran@kinvolk.io>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,431 @@
...
+
+##### Packet
+- Update default machine type from `t1.small.x86` to `c3.small.x86`, since
+    `t1.small.x86` are EOL and no longer available in new Packet projects
+    ([#612](https://github.com/kinvolk/lokomotive/pull/612)).
+
+  If you haven't explicitly defined the controller_type and/or
+  worker_pool.node_type configuration options, upgrading to this release will
+  replace your controller and/or worker nodes with c3.small.x86 machines. To
+  avoid this, set these configuration options to the desired values.
```
Comment (rata) [inline CHANGELOG.md:29]: Are you sure this will happen? I think the cluster will be broken, as terraform will delete the controllers (etcd cluster is broken) and create a new ones, and therefore loose all your data.

Am I missing something?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,431 @@
...
+
+- Add `codespell` target to Makefile and `all` target ([#661](https://github.com/kinvolk/lokomotive/pull/661)).
+
+### Upgrading from v0.2.1
+
+#### Configuration syntax changes
+
+There have been some minor changes to the configurations of following components:
+* contour
+* rometheus-operator.
```
Comment (rata) [inline CHANGELOG.md:129]:
```suggestion
* prometheus-operator.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,431 @@
...
+etcd-0    True      {"health":"true"}
+```
+
+##### Post upgrade steps
+
+##### Openebs
+
+>It is highly recommended to schedule a downtime for the application using the
+>OpenEBS PV while performing this upgrade. Also, make sure you have taken a
+>backup of the data before starting the below upgrade procedure.
```
Comment (rata) [inline CHANGELOG.md:240]: Why the quote? For style/format or comes from openebs docs? If it's the latter, I'd reference that

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,431 @@
+## v0.3.0 - 2020-07-29
+
+We're happy to announce the release of Lokomotive v0.3.0 (Coast Starlight).
+
+This release packs new features and bugfixes. Some of the highlights are:
+
+* Kubernetes 1.18.6
+* AKS Kubernetes 1.16.10
```
Comment (rata) [inline CHANGELOG.md:8]: This seems confusing. Maybe something like:

```suggestion
* Kubernetes 1.18.6
* For Lokomotive clusters running on top of AKS, Kubernetes 1.16.10 is now used
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,431 @@
...
+component "contour" {
+  .
+  .
+}
+
+```
+
+##### Prometheus-operator component
+
+* Variable `external_url` is now removed.
```
Comment (invidian) [inline CHANGELOG.md:159]: Hm, this is actually moved to `ingress.host` field. I think we should mention that to the user. And the valid value changed: now `http://` needs to be removed.

File: CHANGELOG.md (hunk 1 above)
Comment (ipochi) [reply CHANGELOG.md:29]: Yes @rata If the machine type isn't explicity defined, it will amount to a creation of new nodes entirely.
In order to mitigate that existing clusters), the call is to explicitly state in the lokocfg file.

I copy pasted the above paragraph from the `Release Notes` in your PR #612

File: CHANGELOG.md (hunk 3 above)
Comment (invidian) [inline CHANGELOG.md:240]: Hm, that may raise the questions about the backup strategy :smile:

Comment (ipochi) [reply CHANGELOG.md:240]: Would changing to a **NOTE** from a quote help ?

Openebs docs have upgrade procedures for upgrading the control plane components, jiva storage engine and cstore storage engine. I think pointing to the openebs docs would cause more confusion.

File: CHANGELOG.md (hunk 4 above)
Comment (ipochi) [reply CHANGELOG.md:8]:
```suggestion
* Kubernetes 1.18.6
* For Lokomotive clusters running on top of AKS, Kubernetes 1.16.10 is installed.
```

How does that sound ?

Comment (rata) [reply CHANGELOG.md:8]: Perfect! :)

File: CHANGELOG.md (hunk 3 above)
Comment (rata) [reply CHANGELOG.md:240]: I didn't meant pasting the docs, just saying that if this is a quote from openebs docs, add a link to where they come from. Not paste the content, just a link

File: CHANGELOG.md (hunk 1 above)
Comment (rata) [reply CHANGELOG.md:29]: Oh, didn't know it comes from there, thanks!

I still think the release notes on that PR are probably wrong. I think the cluster will be destroyed and all etcd data will be lost if you haven't explicitly defined the controller types.

And for workers it will create downtime too.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,431 @@
...
+- Remove contour ingress workaround(due to an [upstream issue](https://github.com/projectcontour/contour/issues/403))
+  for ExternalDNS ([#635](https://github.com/kinvolk/lokomotive/pull/635)).
+
+#### Development
+
+- Must use consistent build tags for all platforms for CI
+    ([#600](https://github.com/kinvolk/lokomotive/pull/600))
+    & ([#676](https://github.com/kinvolk/lokomotive/pull/676)).
+
+- Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).
```
[the remark this answers is not in the document]
Comment (rata) [reply CHANGELOG.md:97]: What? Why would a user care?

File: CHANGELOG.md (hunk 1 above)
Comment (ipochi) [reply CHANGELOG.md:29]: Yes @rata that's what I said will happen. It will be a new cluster.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,431 @@
...
+    ([#700](https://github.com/kinvolk/lokomotive/pull/700)).
+
+#### Bugfixes
+
+- Remove contour ingress workaround(due to an [upstream issue](https://github.com/projectcontour/contour/issues/403))
+  for ExternalDNS ([#635](https://github.com/kinvolk/lokomotive/pull/635)).
+
+#### Development
+
+- Must use consistent build tags for all platforms for CI
```
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:93]: removed

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,431 @@
...
+    ([#600](https://github.com/kinvolk/lokomotive/pull/600))
+    & ([#676](https://github.com/kinvolk/lokomotive/pull/676)).
+
+- Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).
+
+- Show release values in `--verbose` mode only
```
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:99]: updated.

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:97]: my thoughts exactly ? Why a user would care ?

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,431 @@
...
+    ([#627](https://github.com/kinvolk/lokomotive/pull/627)).
+
+- Change NewClientset to create Kubernetes Clientset from kubeconfig file
```
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:102]: Yes, its not that interesting , but thats why its in the Development section.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,431 @@
...
+content, instead of from file path
+([#628](https://github.com/kinvolk/lokomotive/pull/628)).
+
+- Remove provider aliases from go templates
```
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:106]: Yes, its not that interesting , but thats why its in the Development section.

what others think @iaguis ? I'll remove them if its not needed.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,431 @@
...
+    ([#617](https://github.com/kinvolk/lokomotive/pull/617)).
+
+
+#### Miscellaneous
+
+- Following flatcar-linux/Flatcar#123, Flatcar 2513.1.0 for ARM contains the dig
+    binary so the workaround is no longer needed
+    ([#703](https://github.com/kinvolk/lokomotive/pull/703)).
+
+- Make lint target fail upon invalid git configuration locally. ([#488](https://github.com/kinvolk/lokomotive/pull/488)).
```
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:116]: removed

File: CHANGELOG.md (hunk 5 above)
Comment (ipochi) [reply CHANGELOG.md:159]:
```
Variable `external_url` is now removed and now configured under
    `prometheus.ingress.host`. Remove the http scheme when configuring.
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,431 @@
...
+
+#### Features
+
+- Add `enable_toolbox` variable to [rook-ceph
+component](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/components/rook-ceph)
+([#649](https://github.com/kinvolk/lokomotive/pull/649)). This allows Rook Ceph
+toolbox to manage and configure Ceph.
+
+- Use ingress to expose Prometheus and `external_labels` for federated clusters
+    ([#710](https://github.com/kinvolk/lokomotive/pull/710)).
```
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:61]: added new changes as separate commit, once finalized will squash change long commits into one.

File: CHANGELOG.md (hunk 3 above)
Comment (ipochi) [reply CHANGELOG.md]: Added the link to the openebs documentation.

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (invidian) [reply CHANGELOG.md:101]: According to https://docs.github.com/en/github/administering-a-repository/managing-releases-in-a-repository:
>To notify users that the release is not ready for production and may be unstable, select This is a pre-release.

Previously, all releases have been marked as `pre-release`, which "meant" that they were not ready for production use. Now we do that anymore, which may indicate, that they are actually ready.

However, our motivation is a change in GitHub UI, which if you only have pre-releases, won't show any releases on the main repository page, which is not desired.

I know it's probably a small detail, but in my opinion it can be interpreted as important change.

Alternatively, maybe we don't need to mention that at all, as it has been changed for all releases we have.

[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:101]: If thats the case, then I'd lean on not including it in the release changelog. what do others think ?

File: CHANGELOG.md (hunk 1 above)
Comment (invidian) [reply CHANGELOG.md:29]: We should add a bold warning here, that all cluster data will be lost if user upgrades without specifying it.

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,3 +1,436 @@
...
+    ([#635](https://github.com/kinvolk/lokomotive/pull/635)).
+
+#### Features
+
+- Add `enable_toolbox` variable to [rook-ceph
+component](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/components/rook-ceph)
+([#649](https://github.com/kinvolk/lokomotive/pull/649)). This allows managing
+and configuring Ceph using toolbox pod.
+
+- Add `external_labels` for federated clusters. This helps to identify identify
```
Comment (invidian) [inline CHANGELOG.md:64]: i think we should mention that this is a prometheus feature.

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,3 +1,436 @@
...
+#### Bugfixes
+
+- Remove contour ingress workaround (due to an [upstream issue](https://github.com/projectcontour/contour/issues/403))
+  for ExternalDNS ([#635](https://github.com/kinvolk/lokomotive/pull/635)).
+
+#### Development
+
+- Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).
+
+- Do not show release values in terraform output
```
Comment (invidian) [inline CHANGELOG.md:103]: As it might be confused with project release.
```suggestion
- Do not show Helm release values in terraform output
```

File: CHANGELOG.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (invidian) [reply CHANGELOG.md:106]: I think we should remove it and wait until we can add to the changelog, that asset directory is no longer needed to check cluster health. This is just a tiny and intermediate step towards that.

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,3 +1,436 @@
...
+
+```
+
+##### Prometheus-operator component
+
+* Prometheus specific variables are now under a `prometheus` block.
+* Variable `external_url` is now removed and now configured under
+    `prometheus.ingress.host`. Remove the http scheme when configuring.
+* A new optional `prometheus.ingress` sub-block is introduced to expose
+Prometheus over ingress.
```
Comment (invidian) [inline CHANGELOG.md:165]: Those 2 items should be swapped, to keep it in the logical order, which is:
1. introduction of new block
2. removal of old field (as it has been moved to newly added block).

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,3 +1,436 @@
...
+
+old:
+
+```hcl
+component "prometheus-operator" {
+  .
+  .
+  prometheus_metrics_retention = "14d"
+  prometheus_external_url      = "https://api.example.com/prometheus"
```
Comment (invidian) [inline CHANGELOG.md:174]: We should use the same value for old and new config version, so user can easily map them, as there is no new values introduced here.
```suggestion
  prometheus_external_url      = "https://prometheus.example.lokomotive-k8s.net"
```

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,3 +1,436 @@
...
+    .
+    .
+  }
+  .
+  .
+}
+```
+
+Check out the new syntax in the [Prometheus Operator configuration
+reference](https://github.com/kinvolk/lokomotive/blob/v0.2.0/docs/configuration-reference/components/prometheus-operator.md)
```
Comment (invidian) [inline CHANGELOG.md:211]:
```suggestion
reference](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/components/prometheus-operator.md)
```

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,3 +1,436 @@
...
+>Upgrade functionality is still under active development. It is highly
+>recommended to schedule a downtime for the application using the OpenEBS PV
+>while performing this upgrade. Also, make sure you have taken a backup of the
+>data before starting the below upgrade procedure. - [Openebs
+documentation](https://github.com/openebs/openebs/blob/master/k8s/upgrades/README.md#step-3-upgrade-the-openebs-pools-and-volumes)
+
+
+###### Upgrade cStor Pools
+
+* Extract the SPC name using `kubectl get spc`:
```
Comment (invidian) [inline CHANGELOG.md:255]: We should perhaps open an issue to automate that.

File: CHANGELOG.md [hunk 19]
```diff
@@ -18,7 +451,7 @@ This is a patch release to fix AKS platform deployments.
 
 #### Misc
 
-* Bootkube Docker images are now pulled using Docker protocol, as quay.io plans to deprecate pulling images using ACI ([#656](https://github.com/kinvolk/lokomotive/pull/656).
+* Bootkube Docker images are now pulled using Docker protocol, as quay.io plans to deprecate pulling images using ACI ([#656](https://github.com/kinvolk/lokomotive/pull/656)).
```
Comment (invidian) [inline CHANGELOG.md:454]: This should go into separate commit (and thanks for fixing it!).

File: CHANGELOG.md (hunk 13 above)
Comment (ipochi) [reply CHANGELOG.md:64]:
```suggestion
Add Prometheus feature `external_labels` for federated clusters to Prometheus
    operator component. This helps to identify identify metrics queried from
    different clusters.
```

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:101]: removed.

File: CHANGELOG.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (ipochi) [reply CHANGELOG.md:106]: removed.

File: CHANGELOG.md (hunk 18 above)
Comment (ipochi) [reply CHANGELOG.md:255]: yes, I've already added to the todo list.

Comment (ipochi) [reply CHANGELOG.md:255]: https://github.com/kinvolk/lokomotive/issues/744

File: CHANGELOG.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (invidian) [reply CHANGELOG.md:110]: For correctness:
```suggestion
- Remove Terraform provider aliases from platforms code
```

File: CHANGELOG.md (hunk 1 above)
Comment (ipochi) [reply CHANGELOG.md:29]: added a **WARNING** note.

**Comment by ipochi** [comment]
@invidian addressed all your review comments, please re-review.

File: CHANGELOG.md (hunk 1 above)
Comment (invidian) [reply CHANGELOG.md:29]: Hm, I would rephrase that, to mention, that if not done, you can lose the entire controlplane of your cluster. I'd explain why the type changed, from what to what and include a snippet which should be added if someone do not define machine type at the moment.

Comment (ipochi) [reply CHANGELOG.md:29]: The comment already mentions why the type changed and what changed. I've also added a config snippet.

```
- Update default machine type from `t1.small.x86` to `c3.small.x86`, since
    `t1.small.x86` are EOL and no longer available in new Packet projects
    ([#612](https://github.com/kinvolk/lokomotive/pull/612)).

  **WARNING**: If you haven't explicitly defined the controller_type and/or
  worker_pool.node_type configuration options, upgrading to this release will
  replace your controller and/or worker nodes with c3.small.x86 machines thereby
  loosing all your cluster data. To avoid this, set these configuration options
  to the desired values.

  Make sure that the below variables are explicity defined in your cluster
  configuration. This only applies to machine type `t1.small.x86`.

  ```hcl
  cluster "packet" {
    .
    .
    controller_type = "t1.small.x86"
    .
    .
    worker_pool "pool-name" {
      .
      node_type = "t1.small.x86"
      .
    }
  }
  ```
```

File: CHANGELOG.md [hunk 20]
```diff
@@ -1,3 +1,449 @@
...
+}
+
+```
+
+##### Prometheus-operator component
+
+* Prometheus specific variables are now under a `prometheus` block.
+* A new optional `prometheus.ingress` sub-block is introduced to expose
+Prometheus over ingress.
+* Variable `external_url` is now removed and now configured under
```
Comment (invidian) [inline CHANGELOG.md:177]: As `variable` is:
```hcl
variable "foo" {}
```

```suggestion
* Attribute `external_url` is now removed and now configured under
```

Attribute, because this is how HCL refers to it: https://github.com/hashicorp/hcl/tree/hcl2. (BTW Terraform is inconsistent and calls it `arguments` :facepalm: https://www.terraform.io/docs/configuration/syntax.html)

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,3 +1,449 @@
...
+    `prometheus.ingress.host`. Remove the http scheme when configuring.
```
Comment (invidian) [inline CHANGELOG.md:178]:
```suggestion
    `prometheus.ingress.host`. Remove URL scheme (e.g. `https://`) and URI (e.g. `/prometheus`) when configuring. URI is no longer supported and protocol is always HTTPS.
```

File: CHANGELOG.md [hunk 22]
```diff
@@ -1,3 +1,449 @@
...
+
+old:
+
+```hcl
+component "prometheus-operator" {
+  .
+  .
+  prometheus_metrics_retention = "14d"
+  prometheus_external_url      = "https://prometheus.example.lokomotive-k8s.net""
```
Comment (invidian) [inline CHANGELOG.md:187]:
```suggestion
  prometheus_external_url      = "https://prometheus.example.lokomotive-k8s.net"
```

File: CHANGELOG.md [hunk 23]
```diff
@@ -1,3 +1,449 @@
...
+    metrics_retention = "14d"
+    storage_size      = "50GiB"
+    node_selector = {
+      "kubernetes.io/hostname" = "worker3"
+    }
+
+    ingress {
+      host                       = "prometheus.example.lokomotive-k8s.net"
+      class                      = "contour"
+      certmanager_cluster_issuer = "letsencrypt-production"
```
Comment (invidian) [inline CHANGELOG.md:213]: Are those parameters both required?

Comment (ipochi) [reply CHANGELOG.md:213]: not really, not required.
removing.

File: CHANGELOG.md [hunk 24]
```diff
@@ -1,3 +1,449 @@
+## v0.3.0 - 2020-07-30
```
Comment (rata) [inline CHANGELOG.md:1]: 31 :)

Comment (invidian) [reply CHANGELOG.md:1]: Probably on 08-03 as @ipochi is not around today :smile:

Comment (ipochi) [reply CHANGELOG.md:1]: Today :)

Comment (ipochi) [reply CHANGELOG.md]: done. will need you approvals again.

**Comment by rata** [comment]
More importantly: today :joy:
